### PR TITLE
highligh test failing location on click

### LINKED ISSUE #376: Highlight code in unit testing plugin for failed tests
Investigate and if feasible, pass location with failed tests and highlight code

- [x] add `location` in `remix-tests` result
- [ ] show in UI of SUT Plugin

### PR DESCRIPTION
**Pull request by LianaHus**
fix #376

### CODE REVIEW

**Review by Aniket-Engg — COMMENTED** [review]
- Text highlighting working only after we compile. Do we not get `location` in tests result if contract is not compiled? (I will check that too)

- Highlighting is fine (when it works). Can we also put `red` highlight on line number and gutter same as we do for compiler errors.

File: apps/remix-ide/src/app/tabs/test-tab.js [hunk 1]
```diff
@@ -144,14 +167,23 @@ module.exports = class TestTab extends ViewPlugin {
       this.testsOutput.appendChild(this.outputHeader)
     } else if (result.type === 'testPass') {
       this.testsOutput.appendChild(yo`
-        <div id="${this.runningTestFileName}" data-id="testTabSolidityUnitTestsOutputheader" class="${css.testPass} ${css.testLog} bg-light mb-2 text-success border-0">
+        <div
+          id="${this.runningTestFileName}"
+          data-id="testTabSolidityUnitTestsOutputheader"
+          class="${css.testPass} ${css.testLog} bg-light mb-2 text-success border-0"
+          onclick=${() => this.discardHighlight()}
```
Comment (yann300) [inline apps/remix-ide/src/app/tabs/test-tab.js:174]: why do you `discardHighlight` when one click on a successful test?

**Comment by yann300** [comment]
got this error when clicking on a failed test
```
main.js:9158 Uncaught (in promise) TypeError: Cannot read property 'indexOf' of undefined
    at SourceMappingDecoder.push.../../../dist/libs/remix-debug/src/source/sourceMappingDecoder.js.SourceMappingDecoder.getLinebreakPositions (main.js:9158)
    at OffsetToLineColumnConverter.offsetToLineColumn (main.js:40432)
    at TestTab.highlightLocation (main.js:33615)
    at HTMLDivElement.<anonymous>
```

File: apps/remix-ide/src/app/tabs/test-tab.js [hunk 2]
```diff
@@ -162,7 +194,11 @@ module.exports = class TestTab extends ViewPlugin {
         const method = result.assertMethod === 'ok' ? '' : result.assertMethod
         const expected = result.assertMethod === 'ok' ? `'true'` : result.expected
         this.testsOutput.appendChild(yo`
-          <div class="bg-light mb-2 ${css.testFailure} ${css.testLog} d-flex flex-column text-danger border-0" id="UTContext${result.context}">
+          <div
+            class="bg-light mb-2 ${css.testFailure} ${css.testLog} d-flex flex-column text-danger border-0"
+            id="UTContext${result.context}"
+            onclick=${() => this.highlightLocation(result.location, this.rawFileName)}
```
Comment (yann300) [inline apps/remix-ide/src/app/tabs/test-tab.js:200]: should this be `highlightLocation` or `jumpToLocation` ?

File: apps/remix-ide/src/app/tabs/test-tab.js (hunk 1 above)
Comment (LianaHus) [reply apps/remix-ide/src/app/tabs/test-tab.js:174]: feels better like this. there is no way to remove it, I though to remove it on any click in side panel, but then decided to so this. would you prefer do not remove it?

File: apps/remix-ide/src/app/tabs/test-tab.js (hunk 2 above)
Comment (LianaHus) [reply apps/remix-ide/src/app/tabs/test-tab.js:200]: from test-tab point of view it's more highlightL, form editors point of view it's jump.. but I would say both will work

**Comment by LianaHus** [comment]
> got this error when clicking on a failed test
> 
> ```
> main.js:9158 Uncaught (in promise) TypeError: Cannot read property 'indexOf' of undefined
>     at SourceMappingDecoder.push.../../../dist/libs/remix-debug/src/source/sourceMappingDecoder.js.SourceMappingDecoder.getLinebreakPositions (main.js:9158)
>     at OffsetToLineColumnConverter.offsetToLineColumn (main.js:40432)
>     at TestTab.highlightLocation (main.js:33615)
>     at HTMLDivElement.<anonymous>
> ```

do you remember which kind of error was this?

File: apps/remix-ide/src/app/tabs/test-tab.js (hunk 2 above)
Comment (yann300) [reply apps/remix-ide/src/app/tabs/test-tab.js:200]: is it `highlighting` , i think it doesn't work for me

File: apps/remix-ide/src/lib/offsetToLineColumnConverter.js [hunk 3]
```diff
@@ -36,6 +36,19 @@ export class OffsetToLineColumnConverter extends Plugin {
     return this.sourceMappingDecoder.convertOffsetToLineColumn(rawLocation, this.lineBreakPositionsByContent[file])
   }
 
+  offsetToLineColumnWithContent (rawLocation, file, sources, content) {
```
Comment (yann300) [inline apps/remix-ide/src/lib/offsetToLineColumnConverter.js:39]: it looks like this one is too complex.
`sources` might not really be needed. can be resumed too
```
offsetToLineColumnWithContent (rawLocation, file, content) {
    if (!this.lineBreakPositionsByContent[file]) {
      this.lineBreakPositionsByContent[file] = this.sourceMappingDecoder.getLinebreakPositions(content)
    }
    return this.sourceMappingDecoder.convertOffsetToLineColumn(rawLocation, this.lineBreakPositionsByContent[file])
  }
```

File: apps/remix-ide/src/app/tabs/test-tab.js [hunk 4]
```diff
@@ -144,6 +144,7 @@ module.exports = class TestTab extends ViewPlugin {
         runningTests,
         runningTests[fileName].content
       )
+      this.fileManager.open(fileName)
```
Comment (LianaHus) [inline apps/remix-ide/src/app/tabs/test-tab.js:147]: should be removed

**Review by Aniket-Engg — APPROVED** [review]

**Pull request merged**